fedimint-core/src/transaction.rs

```diff
@@ -125,25 +129,34 @@ pub enum TransactionSignature {
 
 #[derive(Debug, Error, Encodable, Decodable, Clone, Eq, PartialEq)]
 pub enum TransactionError {
-    #[error("The transaction is unbalanced (in={inputs}, out={outputs}, fee={fee})")]
+    #[error("The transaction {txid} is unbalanced (in={inputs}, out={outputs}, fee={fee})")]
     UnbalancedTransaction {
         inputs: Amount,
         outputs: Amount,
         fee: Amount,
+        txid: TransactionId,
     },
-    #[error("The transaction's signature is invalid: tx={tx}, hash={hash}, sig={sig}, key={key}")]
+    #[error("The transaction's {txid} signature is invalid: tx={tx}, sig={sig}, key={key}")]
     InvalidSignature {
+        txid: TransactionId,
         tx: String,
-        hash: String,
         sig: String,
         key: String,
     },
-    #[error("The transaction's signature scheme is not supported: variant={variant}")]
-    UnsupportedSignatureScheme { variant: u64 },
-    #[error("The transaction did not have the correct number of signatures")]
-    InvalidWitnessLength,
-    #[error("The transaction had an invalid input: {}", .0)]
-    Input(DynInputError),
-    #[error("The transaction had an invalid output: {}", .0)]
-    Output(DynOutputError),
+    #[error("The transaction's {txid} signature scheme is not supported: variant={variant}")]
+    UnsupportedSignatureScheme { txid: TransactionId, variant: u64 },
+    #[error("The transaction {txid} did not have the correct number of signatures")]
+    InvalidWitnessLength { txid: TransactionId },
+    #[error("The transaction {txid} had an invalid input at index {}: {}",  .input_idx, .error)]
+    Input {
+        error: DynInputError,
+        input_idx: u64,
+        txid: TransactionId,
+    },
+    #[error("The transaction {txid} had an invalid output at index {}: {}",  .output_idx, .error)]
+    Output {
+        error: DynOutputError,
+        output_idx: u64,
+        txid: TransactionId,
+    },
```

Comment:
Feels like this should be split into a struct containing the txid and an error variant that contains the information currently available. This feels like a useful quick-fix, but not really great code in general.